internal/setup/setup.go

```diff
@@ -30,6 +31,7 @@ type Archive struct {
 	Version    string
 	Suites     []string
 	Components []string
+	Keyrings   []openpgp.KeyRing
```

Comment:
Updated, but I don't think this is proper naming.

IMHO the property in YAML is misnamed because keyring is what it actually holds. It's list of packets that is turned into `openpgp.KeyRing` instance which we use for signature verification and it can contain public keys but also revoked keys.

FWIW it can also contain just a private key, which I use for testing here https://github.com/canonical/chisel/pull/102/files/0dd312696da0c165a5d9c42a2312174e48f663e8#diff-dffb7bab06af83e76768dcafc44e1f9b291acf0b33fd280dc65b970a0bba6ed2R119. I use this keyring to create signatures in `testarchive` and I also pass it to what is now called `PublicKeys`. 

Example:

```
$ gpg --no-default-keyring --keyring /usr/share/keyrings/ubuntu-archive-keyring.gpg -k
/usr/share/keyrings/ubuntu-archive-keyring.gpg
----------------------------------------------
pub   rsa4096 2012-05-11 [SC]
      790BC7277767219C42C86F933B4FE6ACC0B21F32
uid           [ unknown] Ubuntu Archive Automatic Signing Key (2012) <ftpmaster@ubuntu.com>

pub   rsa4096 2012-05-11 [SC]
      843938DF228D22F7B3742BC0D94AA3F0EFE21092
uid           [ unknown] Ubuntu CD Image Automatic Signing Key (2012) <cdimage@ubuntu.com>

pub   rsa4096 2018-09-17 [SC]
      F6ECB3762474EDA9D21B7022871920D1991BC93C
uid           [ unknown] Ubuntu Archive Automatic Signing Key (2018) <ftpmaster@ubuntu.com>

$ gpg --no-default-keyring --keyring /usr/share/keyrings/ubuntu-archive-keyring.gpg --export --armor 
-----BEGIN PGP PUBLIC KEY BLOCK-----

mQINBE+tgXgBEADfiL1KNFHT4H4Dw0OR9LemR8ebsFl+b9E44IpGhgWYDufj0gaM
/UJ1Ti3bHfRT39VVZ6cv1P4mQy0bnAKFbYz/wo+GhzjBWtn6dThYv7n+KL8bptSC
...
E+SWDGxtgwixyPziL56UavL/eeYJWeS/WqvGzZzsAtgSujFVLKWyUaRi0NvYW3h/
I50Tzj0Pkm8GtgvP2UqAWvy+iRpeUQ2ji0Nc
=j6+P
-----END PGP PUBLIC KEY BLOCK-----
```

```
$ gpg --no-default-keyring --keyring /usr/share/keyrings/ubuntu-archive-removed-keys.gpg  -k
/usr/share/keyrings/ubuntu-archive-removed-keys.gpg
---------------------------------------------------
pub   dsa1024 2004-09-12 [SC]
      630239CC130E1A7FD81A27B140976EAF437D05B5
uid           [ unknown] Ubuntu Archive Automatic Signing Key <ftpmaster@ubuntu.com>
sub   elg2048 2004-09-12 [E]

pub   dsa1024 2004-12-30 [SC]
      C5986B4F1257FFA86632CBA746181433FBB75451
uid           [ unknown] Ubuntu CD Image Automatic Signing Key <cdimage@ubuntu.com>

$ gpg --no-default-keyring --keyring /usr/share/keyrings/ubuntu-archive-removed-keys.gpg --export --armor 
-----BEGIN PGP PUBLIC KEY BLOCK-----

mQGiBEFEnz8RBAC7LstGsKD7McXZgd58oN68KquARLBl6rjA2vdhwl77KkPPOr3O
YeSBH/voUsqausJfDNuTNivOfwceDe50lbhq52ODj4Mx9Jg+4aHn9fmRkIk41i2J
...
HcRaDJqO6YL9tsvinm+qciUo3vQVQzF0pptOx49Z0gAHYNU7PULHpvA=
=f4jO
-----END PGP PUBLIC KEY BLOCK-----
```

Reply:
Thanks for the change @woky. In this case, despite both being accurate, I do agree that `PublicKeys` is a better name as it is more explicit about what that variable expects. As @niemeyer said, a Keyring is a container of keys, but in our case, we are restricting that container to only have public keys, and thus it is something much narrower than a keyring. 

Being verbose about it will help with readability.

Reply:
> we are restricting that container to only have public keys

We are not though.

Reply:
Sure. Our point is that the chisel release configuration (yaml) expects a list of public keys, and while a collection of public keys **can** be a keyring, a keyring **might not** be solely composed of public keys

Reply:
> But it doesn't expect this. We specified it as expecting armored dump of keyrings. If we really wanted just a list of public keys, it should have been specified as so.

Thomas, not only we discussed this before but we actually _named the field in the configuration file as public-keys_.  It's not a theory, it was already agreed and you've written down code for this agreement. All that was being suggested is to follow the agreement.

Reply:
> IMHO the property in YAML is misnamed because keyring is what it actually holds. It's list of packets that is turned into openpgp.KeyRing instance which we use for signature verification and it can contain public keys but also revoked keys.

Very well, I don't want us to be wasting more time on this, so let's please do as you imply: let's please change the type and the accepted blocks so that they are actually public keys.

Reply:
FYI I wasn't part of that discussion, but seeing that my arguments are silently ignored, I will just do as you wish to move this forward even though I disagree.

Reply:
> FYI I wasn't part of that discussion, but seeing that my arguments are silently ignored, I will just do as you wish to move this forward even though I disagree.

Yes, you were very much part of this discussion, Thomas, and of the agreement too. Indeed, when we talked you were raising exactly the same issue so it's easy for me to be reminded of it :) and we accepted a compromise that maintained it as a keyring, while labeling it as `public-keys` to make the intent a bit more clear.

But since this is coming for a while already, let's please actually address this and make sure these strings are carrying public key blocks indeed. That's solves your consistency concern, and exposes the intended user interface.

Reply:
Just to be clear, I wrote the spec with the `keyring` property. The spec was ratified in private meeting with you and @cjdcordeiro, and I received it back to implement with the name modified to `public-keys` without any commentary, so I assumed it's just a cosmetic user interface choice. So I really didn't agree nor disagree with anything.

AFAIU, PGP/GPG is designed around keyrings, not individual public keys. If you want to verify a signature against a public key, you first need to create a keyring with only that public key, and verify the signature with that keyring. The set of public keys used to verify signatures and decrypt messages is called _public keyring_ in GPG manual.

Apt uses `gpgv` (called by `apt-key`) which is stripped down version of `gpg` that can only verify signatures, but even this tool operates on keyrings and not individual public keys.

The library we're using for PGP crypto also shares this design. See the public API of the `openpgp` package: https://pkg.go.dev/golang.org/x/crypto/openpgp The only functions to verify signatures take a keyring as an argument. Holding individual instances of `openpgp/packet.PublicKey` instead of `openpgp.Keyring` would require copying some private code from the `openpgp/clearsign` package.

So, yeah, in pure crypto terms, public keys are used to verify signatures, but in GPG, it's the public keyring.

Sorry if you feel that I'm wasting time. I don't mind changing variable/field names and messages to whatever you want, but by changing the data type to what amounts to implementation detail of the library, we're diverging from what's standard and effectively rolling our own crypto with bits from GPG, which is dangerous thing to do.

FYI, I've added a commit that changes the data type anyway so that you see what code we need to add.

Reply:
Thanks for the latest changes @woky. The terminology looks much nicer now.

Wrt the `openpgp` variable types, I don't find it too weird to have `var archivePubKeys openpgp.KeyRing`, but I wonder (maybe naively) why it can't be `var archivePubKeys []openpgp.Key`?

Reply:
> Wrt the openpgp variable types, I don't find it too weird to have var archivePubKeys openpgp.KeyRing, but I wonder (maybe naively) why it can't be var archivePubKeys []openpgp.Key?

@cjdcordeiro I tried to explained it above. In short, in GPG you verify a signature with a public keyring, not with individual public keys. See the issue I mentioned above that would arise if we just blindly tried every key. Does it make sense? By the way, `openpgp.KeyRing` is in fact alias of `[]*openpgp.Entity`.

Reply:
> @niemeyer Just to be clear, I wrote the spec with the keyring property. The spec was ratified in private meeting with you and @cjdcordeiro, and I received it back to implement with the name modified to public-keys without any commentary, so I assumed it's just a cosmetic user interface choice. So I really didn't agree nor disagree with anything.

Not at all. We've had a long conversation about this months ago where you expressed your dissatisfaction with the approach, in exactly the same way you're now extensively doing.

> AFAIU, PGP/GPG is designed around keyrings, not individual public keys.

GPG is designed around keyrings _and_ public keys, quite obviously. Our design will simply accept a `public-keys` map which takes individual public keys as encoded in the typical armored format that is imported and exported by both PGP and GPG. Just run `gpg (...) --export 790BC7277767219C42C86F933B4FE6ACC0B21F32`. That's the content that should go into the field.

> So, yeah, in pure crypto terms, public keys are used to verify signatures, but in GPG, it's the public keyring.
> Sorry if you feel that I'm wasting time.

Honestly, that line of argumentation does feel like a waste of time.

> EDIT: I've started copying bits from the openpgp library

We're not going to fork the openpgp library.

Reply:
> Our design will simply accept a public-keys map which takes individual public keys as encoded in the typical armored format

@niemeyer afaiu that is already the case:

https://github.com/canonical/chisel/blob/a2cf6621a057ea4f028ba639dae3174aa68c796e/internal/setup/setup.go#L320-L334

Reply:
> @niemeyer afaiu that is already the case:

If that was the case we'd not be having this entire thread, right?

Reply:
> We're not going to fork the openpgp library.

But that's what changing the type from `openpgp.KeyRing` to `openpgp/packet.PublicKey` would amount to.

Reply:
We've been through a couple of cases in the past year where this kind of absolutist statement was put to test. So as last time, I'll suggest again: just picture the theory behind this for a moment, Tomas. That would imply that the `openpgp` cannot use the output of `--export <key ID>`. Doesn't that smell a bit awkward?

Reply:
It cannot use `output of --export <key ID>` directly. It needs to import it into keyring and use that keyring to verify signatures. And by the way, `--export <key ID>` does not contain just a public key, it also contains user ID, and sub-keys.

Reply:
Here is some trivial code to process an armored public key:

https://gist.github.com/niemeyer/a4c221dff9fd166500db80d4ce9f817a

I'm closing this PR for now as the conversation is not being productive here.

@cjdcordeiro Let's please coordinate on how to continue this work.

Reply:
Getting a public key from keyring is easy. Using it to verify clearsigned message requires bits from private openpgp code base. And as I said above, trying every key public key against a signature is not the way signatures are supposed to be verified. The keyring abstractions solves this fact. If we used use raw public keys, we'd also need to copy the part that filters those that have given issuer ID matching that of the signature. That's another part of openpgp code we'd need to copy.

Reply:
> Getting a public key from keyring is easy.

Everything is easy, code wise. I have code at hand to verify the signature which is just as simple. But this is not the problem anymore. We need someone that is willing to understanding the underlying rationale, instead of me coding the feature with small incomplete snippets while trying to covey an idea here.

Reply:
Would you mind sharing that code? AFAU, it requires us to copying these blocks of code from the openpgp library:

- https://github.com/ProtonMail/go-crypto/blob/afb1ddc0824c/openpgp/clearsign/clearsign.go#L408-L421
- https://github.com/ProtonMail/go-crypto/blob/afb1ddc0824c/openpgp/clearsign/clearsign.go#L446-L464
- https://github.com/ProtonMail/go-crypto/blob/afb1ddc0824c/openpgp/read.go#L480-L499

IOW, there's no built-in support to verify clearsigned message against a raw public key, we need to copy bits of the library implementation.

I would love to be proven wrong.

Reply:
I've tried to give example implementation using only `packet.PublicKey` another go and it turns out, we'd have to copy more code from the library to be correct. The keys have flags and we also need to filter keys that can sign (verify) messages: https://github.com/ProtonMail/go-crypto/blob/afb1ddc0824ce0052d72ac0d6917f144a1207424/openpgp/keys.go#L338-L366 This information is stored in `openpgp.Key` instance in `SelfSignature` field. So this keyring machinery would need to be copied to `setup.go`.